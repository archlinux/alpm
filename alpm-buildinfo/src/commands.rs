use std::{
    fs::{File, create_dir_all},
    io::{self, IsTerminal, Write},
    str::FromStr,
};

use alpm_common::MetadataFile;
use alpm_types::{SchemaVersion, Sha256Checksum};
use fluent_i18n::t;

use crate::{
    BuildInfo,
    BuildInfoV1,
    BuildInfoV2,
    cli::{CreateCommand, OutputFormat, ValidateArgs},
    error::Error,
};

/// Create a file according to a BUILDINFO schema
pub fn create_file(command: CreateCommand) -> Result<(), Error> {
    let (data, output) = match command {
        CreateCommand::V1 { args } => (
            BuildInfoV1::new(
                args.builddate,
                args.builddir,
                args.buildenv,
                SchemaVersion::from_str("1")?,
                args.installed,
                args.options,
                args.packager,
                args.pkgarch,
                args.pkgbase,
                Sha256Checksum::from_str(&args.pkgbuild_sha256sum)?,
                args.pkgname,
                args.pkgver,
            )?
            .to_string(),
            args.output,
        ),
        CreateCommand::V2 {
            args,
            startdir,
            buildtool,
            buildtoolver,
        } => (
            BuildInfoV2::new(
                args.builddate,
                args.builddir,
                startdir,
                buildtool,
                buildtoolver,
                args.buildenv,
                SchemaVersion::from_str("2")?,
                args.installed,
                args.options,
                args.packager,
                args.pkgarch,
                args.pkgbase,
                Sha256Checksum::from_str(&args.pkgbuild_sha256sum)?,
                args.pkgname,
                args.pkgver,
            )?
            .to_string(),
            args.output,
        ),
    };

    // create any parent directories if necessary
    if let Some(output_dir) = output.0.parent() {
        create_dir_all(output_dir).map_err(|source| Error::IoPathError {
            path: output_dir.to_path_buf(),
            context: t!("error-io-create-output-dir"),
            source,
        })?;
    }

    let mut out = File::create(&output.0).map_err(|source| Error::IoPathError {
        path: output.0.clone(),
        context: t!("error-io-create-output-file"),
        source,
    })?;

    let _ = out
        .write(data.as_bytes())
        .map_err(|source| Error::IoPathError {
            path: output.0,
            context: t!("error-io-write-output-file"),
            source,
        })?;

    Ok(())
}

/// Parses a file according to a BUILDINFO schema.
///
/// Returns a serializable BuildInfo if the file is valid, otherwise an error is returned.
///
/// NOTE: If a command is piped to this process, the input is read from stdin.
/// See [`IsTerminal`] for more information about how terminal detection works.
///
/// [`IsTerminal`]: https://doc.rust-lang.org/stable/std/io/trait.IsTerminal.html
pub fn parse(args: ValidateArgs) -> Result<BuildInfo, Error> {
    if let Some(file) = &args.file {
        BuildInfo::from_file_with_schema(file, args.schema)
    } else if !io::stdin().is_terminal() {
        BuildInfo::from_stdin_with_schema(args.schema)
    } else {
        Err(Error::NoInputFile)
    }
}

/// Validate a file according to a BUILDINFO schema.
///
/// This is a thin wrapper around [`parse`] that only checks if the file is valid.
pub fn validate(args: ValidateArgs) -> Result<(), Error> {
    let _ = parse(args)?;
    Ok(())
}

/// Formats a file according to a BUILDINFO schema.
///
/// Validates and prints the parsed file in the specified output format to stdout.
///
/// The output will be pretty-printed if the `pretty` flag is set to `true` and if the format
/// supports it.
pub fn format(args: ValidateArgs, output_format: OutputFormat, pretty: bool) -> Result<(), Error> {
    let build_info = parse(args)?;
    match output_format {
        OutputFormat::Json => {
            let json = if pretty {
                serde_json::to_string_pretty(&build_info)?
            } else {
                serde_json::to_string(&build_info)?
            };
            println!("{json}");
        }
    }
    Ok(())
}
