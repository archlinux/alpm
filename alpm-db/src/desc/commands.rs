//! Commands for creating, validating, parsing, and formatting [alpm-db-desc] files.
//!
//! [alpm-db-desc]: https://alpm.archlinux.page/specifications/alpm-db-desc.5.html

use std::{
    fs::{File, create_dir_all},
    io::{IsTerminal, Write, stdin},
};

use alpm_common::MetadataFile;

use crate::{
    Error,
    desc::{
        DbDescFile,
        DbDescFileV1,
        DbDescFileV2,
        cli::{CreateCommand, OutputFormat, ValidateArgs},
    },
};

/// Creates an [alpm-db-desc] file according to the specified version command.
///
/// [alpm-db-desc]: https://alpm.archlinux.page/specifications/alpm-db-desc.5.html
///
/// # Notes
///
/// If no output file is specified, the data is printed to stdout.
///
/// # Errors
///
/// Returns an error if:
///
/// - the output file cannot be created,
/// - or the data cannot be written to the output file.
pub fn create_file(command: CreateCommand) -> Result<(), Error> {
    let (data, output) = match command {
        CreateCommand::V1 { args } => {
            let v1 = DbDescFileV1::new(
                args.name,
                args.version,
                args.base,
                args.description,
                args.url,
                args.arch,
                args.builddate,
                args.installdate,
                args.packager,
                args.size,
                args.groups,
                args.reason,
                args.license,
                args.validation,
                args.replaces,
                args.depends,
                args.optdepends,
                args.conflicts,
                args.provides,
            );
            (v1.to_string(), args.output)
        }
        CreateCommand::V2 { args, xdata } => {
            let v2 = DbDescFileV2::new(
                args.name,
                args.version,
                args.base,
                args.description,
                args.url,
                args.arch,
                args.builddate,
                args.installdate,
                args.packager,
                args.size,
                args.groups,
                args.reason,
                args.license,
                args.validation,
                args.replaces,
                args.depends,
                args.optdepends,
                args.conflicts,
                args.provides,
                xdata,
            );
            (v2.to_string(), args.output)
        }
    };

    if let Some(output_path) = output {
        // Create parent directories if necessary
        if let Some(output_dir) = output_path.parent() {
            create_dir_all(output_dir).map_err(|source| Error::IoPathError {
                path: output_dir.to_path_buf(),
                context: "creating output directory",
                source,
            })?;
        }

        let mut out = File::create(&output_path).map_err(|source| Error::IoPathError {
            path: output_path.clone(),
            context: "creating output file",
            source,
        })?;

        out.write_all(data.as_bytes())
            .map_err(|source| Error::IoPathError {
                path: output_path.clone(),
                context: "writing to output file",
                source,
            })?;
    } else {
        print!("{data}");
    }

    Ok(())
}

/// Parses an [alpm-db-desc] file and returns it as a [`DbDescFile`].
///
/// Data can be read from file or stdin.
///
/// # Errors
///
/// Returns an error if
///
/// - valid data cannot be read from a file,
/// - valid data cannot be read from stdin,
/// - or stdin is not a valid file descriptor/TTY.
///
/// [alpm-db-desc]: https://alpm.archlinux.page/specifications/alpm-db-desc.5.html
pub fn parse(args: ValidateArgs) -> Result<DbDescFile, Error> {
    if let Some(file) = &args.file {
        DbDescFile::from_file_with_schema(file, args.schema)
    } else if !stdin().is_terminal() {
        DbDescFile::from_stdin_with_schema(args.schema)
    } else {
        Err(Error::NoInputFile)
    }
}

/// Validate a DB desc file by parsing it.
pub fn validate(args: ValidateArgs) -> Result<(), Error> {
    let _ = parse(args)?;
    Ok(())
}

/// Formats and optionally pretty prints an [alpm-db-desc] file.
///
/// # Errors
///
/// Returns an error if
///
/// - the [alpm-db-desc] file cannot be parsed,
/// - or the data cannot be formatted in the chosen output format.
///
/// [alpm-db-desc]: https://alpm.archlinux.page/specifications/alpm-db-desc.5.html
pub fn format(args: ValidateArgs, output_format: OutputFormat, pretty: bool) -> Result<(), Error> {
    let desc = parse(args)?;
    match output_format {
        OutputFormat::Json => {
            let json = if pretty {
                serde_json::to_string_pretty(&desc).map_err(|e| Error::Json {
                    context: "serializing to pretty JSON",
                    source: e,
                })?
            } else {
                serde_json::to_string(&desc).map_err(|e| Error::Json {
                    context: "serializing to JSON",
                    source: e,
                })?
            };
            println!("{json}");
        }
    }

    Ok(())
}
