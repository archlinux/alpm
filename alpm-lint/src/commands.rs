use std::{env::current_dir, fs::File, io::Write, path::PathBuf};

use alpm_lint::{Error, Level, LintScope, LintStore, Resources, cli::OutputFormat};
use alpm_lint_config::{LintConfiguration, LintGroup, LintRuleConfiguration};
use log::debug;
use serde::Serialize;
use strum::VariantArray;

/// Writes `content` to stdout or a file.
///
/// If `output_path` is a file path, `content` is written to it, else `content` is written to
/// stdout.
///
/// # Errors
///
/// Returns an error if an output file cannot be created or written to.
fn write_output(content: &str, output_path: Option<PathBuf>) -> Result<(), Error> {
    match output_path {
        Some(path) => {
            let mut file = File::create(&path).map_err(|source| Error::IoPath {
                path: path.clone(),
                context: "creating output file",
                source,
            })?;
            file.write_all(content.as_bytes())
                .map_err(|source| Error::IoPath {
                    path,
                    context: "writing to output file",
                    source,
                })?;
        }
        None => print!("{content}"),
    }
    Ok(())
}

/// Takes any serializable object and serializes it into the given [`OutputFormat`].
///
/// # Errors
///
/// Returns an error if serialization fails.
fn serialize_output<T: Serialize>(
    object: T,
    output_format: OutputFormat,
    pretty: bool,
    context: &str,
) -> Result<String, Error> {
    let output = match output_format {
        OutputFormat::Json => {
            if pretty {
                serde_json::to_string_pretty(&object).map_err(|error| Error::Json {
                    error,
                    context: context.into(),
                })?
            } else {
                serde_json::to_string(&object).map_err(|error| Error::Json {
                    error,
                    context: context.into(),
                })?
            }
        }
    };

    Ok(output)
}

/// Runs a lint check.
///
/// If not provided, the `path` and `scope` are automatically detected.
/// Defaults to the current working directory if no `path` is provided.
pub fn check(
    config_path: Option<PathBuf>,
    path: Option<PathBuf>,
    scope: Option<LintScope>,
    level: Level,
    format: LintOutputFormat,
    output: Option<PathBuf>,
    pretty: bool,
) -> Result<(), Error> {
    let path = match path {
        Some(path) => path,
        None => current_dir().map_err(|source| Error::Io {
            context: "detect current working directory",
            source,
        })?,
    };
    debug!("Using path: {path:?}");

    // Load the config or fall back to the default config.
    let config = if let Some(path) = config_path {
        LintConfiguration::from_path(&path)?
    } else {
        LintConfiguration::default()
    };

    // Get or detect the scope.
    let scope = match scope {
        Some(scope) => scope,
        None => LintScope::detect(&path)?,
    };
    debug!("Detected scope: {scope:?}");

    let resources = Resources::gather(&path, scope)?;
    debug!("Resources have been gathered.");

    let store = LintStore::new(config);

    let mut issues = Vec::new();
    let lint_rules = store.filtered_lint_rules(&scope, level);

    debug!("Start of linting.");
    for (name, rule) in lint_rules {
        debug!("Running rule: '{name}'");
        rule.run(&resources, &mut issues)?;
    }

    let found_issues = !issues.is_empty();

    debug!("Using output format {format:?}.");
    let content = match format {
        None => unimplemented!(),
        Some(output_format) => serialize_output(issues, output_format, pretty)?,
    };

    write_output(&content, output)?;

    // Exit with code 1 if there were any relevant lints.
    if found_issues {
        std::process::exit(1);
    }

    Ok(())
}

/// Writes the definition of all linting rules to output.
///
/// Writes a map of [`SerializableLintRule`]s in serialized form.
pub fn rules(
    output_format: OutputFormat,
    pretty: bool,
    output: Option<PathBuf>,
) -> Result<(), Error> {
    // Create a stub default configuration.
    let config = LintConfiguration::default();
    let store = LintStore::new(config);

    let lint_rules = store.serializable_lint_rules();
    let content = serialize_output(lint_rules, output_format, pretty, "lint rules")?;

    write_output(&content, output)?;
    Ok(())
}

/// Writes the definitions of all linting options to output.
pub fn options(
    output_format: OutputFormat,
    pretty: bool,
    output: Option<PathBuf>,
) -> Result<(), Error> {
    let options = LintRuleConfiguration::configuration_options();
    let content = serialize_output(options, output_format, pretty, "lint options")?;

    write_output(&content, output)?;
    Ok(())
}

/// Metadata information for static site generator integration.
///
/// Contains all available lint groups, scopes, and levels that can be used
/// to create dropdown fields in a static site generator.
#[derive(Debug, Serialize)]
pub struct Meta {
    /// All available lint groups.
    pub groups: Vec<LintGroup>,
    /// All available lint scopes.
    pub scopes: Vec<LintScope>,
    /// All available lint levels.
    pub levels: Vec<Level>,
}

impl Default for Meta {
    fn default() -> Self {
        Self::new()
    }
}

impl Meta {
    /// Creates a new [`Meta`] instance with all available groups, scopes, and levels.
    ///
    /// This function collects all enum variants for lint groups, scopes, and levels
    /// that can be used in configuration and CLI options.
    pub fn new() -> Self {
        Self {
            groups: LintGroup::VARIANTS.to_vec(),
            scopes: LintScope::VARIANTS.to_vec(),
            levels: Level::VARIANTS.to_vec(),
        }
    }
}

/// Writes metadata about available lint groups, scopes, and levels to output.
///
/// This is primarily intended for static site generators that need to create
/// dropdown fields based on the available configuration options.
pub fn meta(
    output_format: OutputFormat,
    pretty: bool,
    output: Option<PathBuf>,
) -> Result<(), Error> {
    let meta = Meta::new();
    let content = serialize_output(meta, output_format, pretty, "lint meta types")?;

    write_output(&content, output)?;
    Ok(())
}
