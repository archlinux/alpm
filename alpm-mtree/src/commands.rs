use std::{
    io::{self, IsTerminal},
    path::PathBuf,
};

use alpm_common::MetadataFile;

use crate::{Error, Mtree, MtreeSchema, cli::OutputFormat};

/// A small wrapper around the parsing of an MTREE file that simply ensures that there were no
/// errors.
///
/// For all possible errors, check the [parse] function.
pub fn validate(file: Option<&PathBuf>, schema: Option<MtreeSchema>) -> Result<(), Error> {
    parse(file, schema)?;

    Ok(())
}

/// Parse a given file and output it in the specified format to stdout.
///
/// # Errors
///
/// Returns an error if the input can not be parsed and validated, or if the output can not be
/// formatted in the selected output format.
pub fn format(
    file: Option<&PathBuf>,
    schema: Option<MtreeSchema>,
    format: OutputFormat,
    pretty: bool,
) -> Result<(), Error> {
    let files = parse(file, schema)?;

    match format {
        OutputFormat::Json => {
            let json = if pretty {
                serde_json::to_string_pretty(&files)?
            } else {
                serde_json::to_string(&files)?
            };
            println!("{json}");
        }
    }

    Ok(())
}

/// Parse and interpret an MTREE file.
///
/// 1. Reads the contents of a file or stdin.
/// 2. Check whether the input is gzip compressed as that's how it's delivered inside of packages.
/// 3. Parse the input
///
/// NOTE: If a command is piped to this process, the input is read from stdin.
/// See [`IsTerminal`] for more information about how terminal detection works.
///
/// [`IsTerminal`]: https://doc.rust-lang.org/stable/std/io/trait.IsTerminal.html
///
/// # Errors
///
/// - [Error::NoInputFile] if a file is given and doesn't exist.
/// - [Error::IoPath] if a given file cannot be opened or read.
/// - [Error::Io] if the file is streamed via StdIn and an error occurs.
/// - [Error::InvalidGzip] if the file is gzip compressed, but the archive is malformed.
/// - [Error::InvalidUTF8] if the given file contains invalid UTF-8.
/// - [Error::ParseError] if a malformed MTREE file is encountered.
/// - [Error::InterpreterError] if expected properties for a given type aren't set.
pub fn parse(file: Option<&PathBuf>, schema: Option<MtreeSchema>) -> Result<Mtree, Error> {
    if let Some(file) = file {
        Mtree::from_file_with_schema(file, schema)
    } else if !io::stdin().is_terminal() {
        Mtree::from_stdin_with_schema(schema)
    } else {
        Err(Error::NoInputFile)
    }
}
