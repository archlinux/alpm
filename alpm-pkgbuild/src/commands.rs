//! Definition of the high-level binary entry points.

use std::path::PathBuf;

use alpm_common::MetadataFile;
use alpm_pkgbuild::{
    bridge::{BridgeOutput, run_bridge_script},
    cli::OutputFormat,
    error::Error,
};
use alpm_srcinfo::{SourceInfo, SourceInfoV1};

/// Run the bridge script on a `PKGBUILD` and return the output.
///
/// # Errors
///
/// Returns an error if `run_bridge_script` fails.
pub fn run_bridge(pkgbuild_path: PathBuf) -> Result<(), Error> {
    println!("{}", run_bridge_script(&pkgbuild_path)?);

    Ok(())
}

/// Takes a [PKGBUILD], creates [SRCINFO] data from it and prints it.
///
/// # Errors
///
/// Returns an error if
///
/// - running the `alpm-pkgbuild-bridge` script fails,
/// - or parsing the output of the `alpm-pkgbuild-bridge` script fails.
///
/// [PKGBUILD]: https://man.archlinux.org/man/PKGBUILD.5
/// [SRCINFO]: https://alpm.archlinux.page/specifications/SRCINFO.5.html
pub fn print_source_info(
    pkgbuild_path: PathBuf,
    output_format: OutputFormat,
    pretty: bool,
) -> Result<(), Error> {
    let output = BridgeOutput::from_file(&pkgbuild_path)?;
    let source_info: SourceInfoV1 = output.try_into()?;

    match output_format {
        OutputFormat::Json => {
            let json = if pretty {
                serde_json::to_string_pretty(&source_info)?
            } else {
                serde_json::to_string(&source_info)?
            };
            println!("{json}");
        }
        OutputFormat::Srcinfo => {
            println!("{}", source_info.as_srcinfo())
        }
    }

    Ok(())
}

/// Runs the `alpm-pkgbuild-bridge` script on a [PKGBUILD] and returns the output.
///
/// This is a development and debug command to better understand the inner workings of this binary.
///
/// If the generated and read [SRCINFO] representations do not match, the respective files
/// `pkgbuild.json` and `srcinfo.json` are output to the current working directory and the function
/// exits with an exit code of `1`.
///
/// These files contain pretty-printed JSON, which accurately depicts the internal representation
/// used to compare the two files.
///
/// # Errors
///
/// Returns an error if
///
/// - running the `alpm-pkgbuild-bridge` script fails,
/// - creating a [`SrcInfoV1`] from the script output fails,
/// - creating a [`SrcInfo`] from the the [SRCINFO] file fails,
/// - or creating JSON representations for either [SRCINFO] data fails in case of mismatch.
///
/// [PKGBUILD]: https://man.archlinux.org/man/PKGBUILD.5
/// [SRCINFO]: https://alpm.archlinux.page/specifications/SRCINFO.5.html
pub fn compare_source_info(pkgbuild_path: PathBuf, srcinfo_path: PathBuf) -> Result<(), Error> {
    let output = BridgeOutput::from_file(&pkgbuild_path)?;
    let pkgbuild_source_info: SourceInfoV1 = output.try_into()?;

    let source_info = SourceInfo::from_file_with_schema(srcinfo_path, None)?;
    let SourceInfo::V1(source_info) = source_info;

    if source_info != pkgbuild_source_info {
        let pkgbuild_source_info = serde_json::to_string_pretty(&pkgbuild_source_info)?;
        let source_info = serde_json::to_string_pretty(&source_info)?;

        let pkgbuild_json_path = PathBuf::from("pkgbuild.json");
        std::fs::write("pkgbuild.json", pkgbuild_source_info).map_err(|source| Error::IoPath {
            path: pkgbuild_json_path,
            context: "writing pkgbuild.json file",
            source,
        })?;
        let srcinfo_json_path = PathBuf::from("srcinfo.json");
        std::fs::write("srcinfo.json", source_info).map_err(|source| Error::IoPath {
            path: srcinfo_json_path,
            context: "writing srcinfo.json file",
            source,
        })?;

        eprintln!(
            "SRCINFO data generated from PKGBUILD file differs from the .SRCINFO file read from disk.\n\
            Compare the two generated files pkgbuild.json and srcinfo.json for details."
        );
        std::process::exit(1);
    } else {
        println!("The generated content matches that read from disk.");
    }

    Ok(())
}
