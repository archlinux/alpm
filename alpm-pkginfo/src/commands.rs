//! Command-line functions, that are called by the `alpm-pkginfo` executable.

use std::{
    fs::{File, create_dir_all},
    io::{self, IsTerminal, Write},
    path::PathBuf,
};

use alpm_common::MetadataFile;
use fluent_i18n::t;

use crate::{
    Error,
    PackageInfo,
    PackageInfoSchema,
    PackageInfoV1,
    PackageInfoV2,
    cli::{CreateCommand, OutputFormat},
};

/// Create a file according to a PKGINFO schema
///
/// ## Errors
///
/// Returns an error if one of the provided arguments is not valid, if creating output directory or
/// file is not possible, or if the output file can not be written to.
pub fn create_file(command: CreateCommand) -> Result<(), Error> {
    let (data, output) = match command {
        CreateCommand::V1 { args } => (
            PackageInfoV1::new(
                args.pkgname,
                args.pkgbase,
                args.pkgver,
                args.pkgdesc,
                args.url,
                args.builddate,
                args.packager,
                args.size,
                args.arch,
                args.license,
                args.replaces,
                args.group,
                args.conflict,
                args.provides,
                args.backup,
                args.depend,
                args.optdepend,
                args.makedepend,
                args.checkdepend,
            )
            .to_string(),
            args.output,
        ),
        CreateCommand::V2 { args, xdata } => (
            PackageInfoV2::new(
                args.pkgname,
                args.pkgbase,
                args.pkgver,
                args.pkgdesc,
                args.url,
                args.builddate,
                args.packager,
                args.size,
                args.arch,
                args.license,
                args.replaces,
                args.group,
                args.conflict,
                args.provides,
                args.backup,
                args.depend,
                args.optdepend,
                args.makedepend,
                args.checkdepend,
                xdata,
            )?
            .to_string(),
            args.output,
        ),
    };

    // create any parent directories if necessary
    if let Some(output_dir) = output.0.parent() {
        create_dir_all(output_dir).map_err(|source| Error::IoPathError {
            path: output_dir.to_path_buf(),
            context: t!("error-io-create-output-dir"),
            source,
        })?;
    }

    let mut out = File::create(&output.0).map_err(|source| Error::IoPathError {
        path: output.0.clone(),
        context: t!("error-io-create-output-file"),
        source,
    })?;

    let _ = out
        .write(data.as_bytes())
        .map_err(|source| Error::IoPathError {
            path: output.0,
            context: t!("error-io-write-output-file"),
            source,
        })?;

    Ok(())
}

/// Parses a file according to a PKGINFO schema.
///
/// Returns a serializable PackageInfo if the file is valid, otherwise an error is returned.
///
/// NOTE: If a command is piped to this process, the input is read from stdin.
/// See [`IsTerminal`] for more information about how terminal detection works.
///
/// [`IsTerminal`]: https://doc.rust-lang.org/stable/std/io/trait.IsTerminal.html
///
/// ## Errors
///
/// Returns an error if the file can not be read, if the file can not be parsed, or if the file is
/// not valid according to the schema.
pub fn parse(
    file: Option<PathBuf>,
    schema: Option<PackageInfoSchema>,
) -> Result<PackageInfo, Error> {
    if let Some(file) = file {
        PackageInfo::from_file_with_schema(file, schema)
    } else if !io::stdin().is_terminal() {
        PackageInfo::from_stdin_with_schema(schema)
    } else {
        Err(Error::NoInputFile)
    }
}

/// Validate a file according to a PKGINFO schema.
///
/// This is a thin wrapper around [`parse`] that only checks if the file is valid.
///
/// ## Errors
///
/// Returns an error if parsing `file` fails.
pub fn validate(file: Option<PathBuf>, schema: Option<PackageInfoSchema>) -> Result<(), Error> {
    let _ = parse(file, schema)?;
    Ok(())
}

/// Formats a file according to a PKGINFO schema.
///
/// Validates and prints the parsed file in the specified output format to stdout.
///
/// The output will be pretty-printed if the `pretty` flag is set to `true` and if the format
/// supports it.
///
/// ## Errors
///
/// Returns an error if parsing of `file` fails or if the output format can not be created.
pub fn format(
    file: Option<PathBuf>,
    schema: Option<PackageInfoSchema>,
    output_format: OutputFormat,
    pretty: bool,
) -> Result<(), Error> {
    let pkg_info = parse(file, schema)?;
    match output_format {
        OutputFormat::Json => {
            println!(
                "{}",
                if pretty {
                    serde_json::to_string_pretty(&pkg_info)?
                } else {
                    serde_json::to_string(&pkg_info)?
                }
            );
        }
    }
    Ok(())
}
