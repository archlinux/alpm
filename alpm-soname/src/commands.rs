//! Command line functions that are called by the `alpm-soname` executable.

use std::io::Write;

use alpm_types::{Soname, SonameLookupDirectory};

use crate::{
    Error,
    cli::{OutputFormat, PackageArgs},
    find_dependencies,
    find_provisions,
    lookup::extract_elf_sonames,
};

/// Get the provisions of a package and print them to the given output.
///
/// See the [`find_provisions`] function for more details.
///
/// # Errors
///
/// Returns an error if [`find_provisions`] returns an error or if the output stream
/// can not be written to.
pub fn get_provisions<W: Write>(
    args: PackageArgs,
    lookup_dir: SonameLookupDirectory,
    output: &mut W,
) -> Result<(), Error> {
    let provisions = find_provisions(args.package, lookup_dir)?;

    match args.output_format {
        OutputFormat::Plain => {
            for provision in provisions {
                writeln!(output, "{provision}").map_err(|source| Error::IoWriteError {
                    context: "writing provision to output",
                    source,
                })?;
            }
        }
        OutputFormat::Json => {
            let json = if args.pretty {
                serde_json::to_string_pretty(&provisions)?
            } else {
                serde_json::to_string(&provisions)?
            };
            writeln!(output, "{json}").map_err(|source| Error::IoWriteError {
                context: "writing JSON to output",
                source,
            })?;
            return Ok(());
        }
    }

    Ok(())
}

/// Get the dependencies of a package and print them to the given output.
///
/// See the [`find_dependencies`] functions for more details.
///
/// # Errors
///
/// Returns an error if [`find_dependencies`] returns an error or if the output stream
/// can not be written to.
pub fn get_dependencies<W: Write>(
    args: PackageArgs,
    lookup_dir: SonameLookupDirectory,
    output: &mut W,
) -> Result<(), Error> {
    let dependencies = find_dependencies(args.package, lookup_dir)?;

    match args.output_format {
        OutputFormat::Plain => {
            for dependency in &dependencies {
                writeln!(output, "{dependency}").map_err(|source| Error::IoWriteError {
                    context: "writing dependency to output",
                    source,
                })?;
            }
            return Ok(());
        }
        OutputFormat::Json => {
            let json = serde_json::to_string_pretty(&dependencies)?;
            writeln!(output, "{json}").map_err(|source| Error::IoWriteError {
                context: "writing JSON to output",
                source,
            })?;
        }
    }

    Ok(())
}

/// Get the raw ELF soname dependencies of a package and print them to the given output.
///
///
/// Unlike [`get_dependencies`], this function does not filter the dependencies by the lookup
/// directory. In other words, it prints all ELF sonames found in the package regardless of whether
/// they match the lookup directory or not.
///
/// See the [`extract_elf_sonames`] function for more details.
///
/// # Errors
///
/// Returns an error if [`extract_elf_sonames`] returns an error or if the output stream
/// can not be written to.
pub fn get_raw_dependencies<W: Write>(args: PackageArgs, output: &mut W) -> Result<(), Error> {
    let mut elf_sonames: Vec<Soname> = extract_elf_sonames(args.package)?
        .into_iter()
        .flat_map(|elf_soname| elf_soname.sonames)
        .collect();
    elf_sonames.sort();
    elf_sonames.dedup();

    match args.output_format {
        OutputFormat::Plain => {
            for elf_soname in elf_sonames {
                writeln!(output, "{elf_soname}").map_err(|source| Error::IoWriteError {
                    context: "writing ELF soname to output",
                    source,
                })?;
            }
        }
        OutputFormat::Json => {
            let json = if args.pretty {
                serde_json::to_string_pretty(&elf_sonames)?
            } else {
                serde_json::to_string(&elf_sonames)?
            };
            writeln!(output, "{json}").map_err(|source| Error::IoWriteError {
                context: "writing JSON to output",
                source,
            })?;
        }
    }

    Ok(())
}
