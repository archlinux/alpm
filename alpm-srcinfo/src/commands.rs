//! Functions called from the binary.
use std::{
    fs::File,
    io::{self, BufReader, IsTerminal, Read},
    path::PathBuf,
};

use winnow::Parser;

use crate::{error::Error, parser::SourceInfoContent};

/// Validates a SRCINFO file from a path or stdin.
///
/// Wraps the [`parse`] function and allows to ensure that no errors occurred during parsing.
pub fn validate(file: Option<&PathBuf>) -> Result<(), Error> {
    parse(file)?;

    Ok(())
}

/// Parses and interprets a SRCINFO file from a path or stdin.
///
/// ## Note
///
/// If a command is piped to this process, the input is read from stdin.
/// See [`IsTerminal`] for more information about how terminal detection works.
///
/// [`IsTerminal`]: https://doc.rust-lang.org/stable/std/io/trait.IsTerminal.html
///
/// # Errors
///
/// Returns an error if the input can not be parsed and validated, or if the output can not be
/// formatted in the selected output format.
pub fn parse(file: Option<&PathBuf>) -> Result<(), Error> {
    // The buffer that'll contain the raw file.
    let mut buffer = Vec::new();

    // Read the file into the buffer, either from a given file or stdin.
    if let Some(path) = file {
        let file =
            File::open(path).map_err(|err| Error::IoPath(path.clone(), "opening file", err))?;
        let mut buf_reader = BufReader::new(file);
        buf_reader
            .read_to_end(&mut buffer)
            .map_err(|err| Error::IoPath(path.clone(), "reading file", err))?;
    } else if !io::stdin().is_terminal() {
        let mut buffer = Vec::new();
        let mut stdin = io::stdin();
        stdin
            .read_to_end(&mut buffer)
            .map_err(|err| Error::Io("reading from stdin", err))?;
    } else {
        return Err(Error::NoInputFile);
    };

    let contents = String::from_utf8(buffer)?.to_string();

    // Parse the given srcinfo file.
    SourceInfoContent::parser
        .parse(&contents)
        .map_err(|err| Error::ParseError(format!("{err}")))?;

    Ok(())
}
