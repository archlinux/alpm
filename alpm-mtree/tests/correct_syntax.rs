//! Syntax integration tests for the `alpm-mtree` CLI.

#![cfg(feature = "cli")]

use std::{fs::read_to_string, path::PathBuf};

use alpm_mtree::commands::parse;
use insta::assert_snapshot;
use rstest::rstest;
use testresult::TestResult;

/// Happy path testing for mtree.
///
/// Take some input, parse it and compare the serialized JSON output with the snapshot.
#[rstest]
fn ensure_correct_syntax(#[files("tests/correct_syntax_inputs/*")] case: PathBuf) -> TestResult {
    // Read the input file and parse it.

    let input = read_to_string(&case)?;
    let result = parse(Some(&case), None);

    // Make sure the parsing succeeded.
    let files = match result {
        Ok(parsed) => parsed,
        Err(error) => {
            eprintln!("The parser failed even though it should've succeeded for input:\n{input}");
            return Err(format!("{error}").into());
        }
    };

    let name = case.file_stem().unwrap().to_str().unwrap();

    let pretty_json = serde_json::to_string_pretty(&files)?;

    // Run the tests with the input being displayed as the description.
    // This makes reviewing this whole stuff a lot easier.
    // Also remove the usual module prefix by explicitly setting the snapshot path.
    // This isn't necessary, as we're already manually sorting snapshots by test scenario.
    let input_clone = input.clone();
    insta::with_settings!({
        description => input_clone,
        snapshot_path => "correct_syntax_snapshots",
        prepend_module_to_snapshot => false,
    }, {
        assert_snapshot!(name, pretty_json);
    });

    Ok(())
}
