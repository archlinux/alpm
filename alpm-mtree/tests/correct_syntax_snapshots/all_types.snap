---
source: alpm-mtree/tests/correct_syntax.rs
description: "#mtree\n/set mode=644 uid=0 gid=0 type=file\n./some_file time=1700000000.0 size=1337 sha256digest=0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef\n./some_link type=link link=some_file time=1700000000.0\n./some_dir type=dir time=1700000000.0\n"
expression: pretty_json
---
[
  {
    "type": "dir",
    "path": "./some_dir",
    "uid": 0,
    "gid": 0,
    "mode": "644",
    "time": 1700000000
  },
  {
    "type": "file",
    "path": "./some_file",
    "uid": 0,
    "gid": 0,
    "mode": "644",
    "size": 1337,
    "time": 1700000000,
    "sha256_digest": "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef"
  },
  {
    "type": "link",
    "path": "./some_link",
    "uid": 0,
    "gid": 0,
    "mode": "644",
    "time": 1700000000,
    "link_path": "some_file"
  }
]
